test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlQueueClientTests.cs

```diff
@@ -322,36 +328,42 @@ public async Task ExecuteWithHeartbeats()
             cancel.CancelAfter(TimeSpan.FromSeconds(30));
             var execDate = DateTime.UtcNow;
             var dequeueDate = DateTime.UtcNow;
-            var execTask = JobHosting.ExecuteJobWithHeartbeatsAsync(
-                client,
-                queueType,
-                job.Id,
-                job.Version,
-                async cancel =>
-                {
-                    await Task.Delay(TimeSpan.FromSeconds(10));
-                    execDate = DateTime.UtcNow;
-                    return "Test";
-                },
-                TimeSpan.FromSeconds(1),
-                cancel);
-            var jobInt = (JobInfo)null;
-            var dequeueTask = Task.Run(
-                async () =>
-                {
-                    while (jobInt == null)
-                    {
-                        await Task.Delay(TimeSpan.FromSeconds(1));
-                        jobInt = await client.DequeueAsync(queueType, "test-worker", 2, cancel.Token);
-                    }
 
-                    dequeueDate = DateTime.UtcNow;
+            this.Retry(
+                () =>
+                {
+                    var execTask = JobHosting.ExecuteJobWithHeartbeatsAsync(
+                            client,
+                            queueType,
+                            job.Id,
+                            job.Version,
+                            async cancel =>
+                            {
+                                await Task.Delay(TimeSpan.FromSeconds(10));
+                                execDate = DateTime.UtcNow;
+                                return "Test";
+                            },
+                            TimeSpan.FromSeconds(1),
+                            cancel);
+                    var jobInt = (JobInfo)null;
+                    var dequeueTask = Task.Run(
+                        async () =>
+                        {
+                            while (jobInt == null)
+                            {
+                                await Task.Delay(TimeSpan.FromSeconds(1));
+                                jobInt = await client.DequeueAsync(queueType, "test-worker", 2, cancel.Token);
+                            }
+
+                            dequeueDate = DateTime.UtcNow;
+                        },
+                        cancel.Token);
+                    Task.WaitAll(execTask, dequeueTask);
+
+                    Assert.Equal(job.Id, jobInt.Id);
+                    Assert.True(dequeueDate >= execDate, $"dequeue:{dequeueDate} >= exec:{execDate}");
                 },
-                cancel.Token);
-            Task.WaitAll(execTask, dequeueTask);
-
-            Assert.Equal(job.Id, jobInt.Id);
-            Assert.True(dequeueDate >= execDate, $"dequeue:{dequeueDate} >= exec:{execDate}");
+                additionalRetriableExceptions: _retriableExceptions);
```

Comment:
I think this test does highlight that the underlying code doesn't handle this condition very well. In the above, a Heartbeat event can be written while the Dequeue operation is processing causing a Pre-condition failure. I'm wondering if we should just build this logic into Dequeue to retry when this this happens. Thoughts?

Reply:
Yeap. I see your suggestion, and I think it makes total sense. 
And I don't see places catching the RetriableJobException raised by Dequeue.

Reply:
Added a logic to retry retriable exceptions up to three times, unless the exception is not retriable.